h2o-core/src/main/java/water/persist/PersistManager.java

```diff
@@ -258,24 +260,30 @@ public void importFiles(String path, ArrayList<String> files, ArrayList<String>
       } catch( Throwable e) {
         fails.add(path); // Fails for e.g. broken sockets silently swallow exceptions and just record the failed path
       }
-      return;
     }
-
-    if(s.startsWith("s3:")) {
+    else if(s.startsWith("s3:")) {
       if (I[Value.S3] == null) throw new H2OIllegalArgumentException("S3 support is not configured");
-      I[Value.S3].importFiles(path, files, keys, fails, dels);
-      return;
+      I[Value.S3].importFiles(path, pattern, files, keys, fails, dels);
     }
-    if( s.startsWith("hdfs:") ||
+    else if( s.startsWith("hdfs:") ||
         s.startsWith("s3n:") || 
         s.startsWith("s3a:") || 
         s.startsWith("maprfs:")) {
       if (I[Value.HDFS] == null) throw new H2OIllegalArgumentException("HDFS, S3N, and S3A support is not configured");
-      I[Value.HDFS].importFiles(path, files, keys, fails, dels);
-      return;
+      I[Value.HDFS].importFiles(path, pattern, files, keys, fails, dels);
+    }
+    else {
+      // Attempt NFS import instead
+      I[Value.NFS].importFiles(path, pattern, files, keys, fails, dels);
     }
-    // Attempt NFS import instead
-    I[Value.NFS].importFiles(path, files, keys, fails, dels);
+    if(!pattern.isEmpty()) {
+      ArrayList<String> subsetFiles = matchPattern(path,files,pattern); //New files ArrayList after matching pattern of choice
+      files.retainAll(subsetFiles); //Subset files based on subsetFiles
+      ArrayList<String> subsetKey = matchPattern(path,keys,pattern); //New keys ArrayList after matching pattern of choice
```

Comment:
I do not understand the matching process here. Why `matchPattern` is called twice? You need to select only files, and then post-process keys and select only keys corresponding to selected files.

Furthermore, you should also filter `fails` since we should report failure only for files described by user pattern not all.